substrate/frame/staking/src/ledger.rs

```diff
@@ -106,11 +106,31 @@ impl<T: Config> StakingLedger<T> {
 	/// This getter can be called with either a controller or stash account, provided that the
 	/// account is properly wrapped in the respective [`StakingAccount`] variant. This is meant to
 	/// abstract the concept of controller/stash accounts from the caller.
+	///
+	/// Returns [`Error::BadState`] when a bond is in "bad state". A bond is in a bad state when a
+	/// stash has a controller which is bonding a ledger associated with another stash.
 	pub(crate) fn get(account: StakingAccount<T::AccountId>) -> Result<StakingLedger<T>, Error<T>> {
-		let controller = match account {
-			StakingAccount::Stash(stash) => <Bonded<T>>::get(stash).ok_or(Error::<T>::NotStash),
-			StakingAccount::Controller(controller) => Ok(controller),
-		}?;
+		let (stash, controller) = match account.clone() {
+			StakingAccount::Stash(stash) =>
+				(stash.clone(), <Bonded<T>>::get(&stash).ok_or(Error::<T>::NotStash)?),
```

Comment:
```suggestion
				(stash.clone(), <Bonded<T>>::get(&stash).defensive_ok_or(Error::<T>::NotStash)?),
```

Reply:
Should this be defensive? The `StakingLedger::get()` is also used in extrinsics like validate, nominate, etc so we may be throwing defensives on user calls, which are expected from a runtime perspective if the user enters the wrong account IDs.

Reply:
I suggested it to be defensive because if the caller of this function uses `StakingAccount::Stash(_)`, then `Bonded` storage item _MUST_ exist, and similarly for `StakingAccount::Controller(_)`.

But I see what you mean; if someone calls `staking.validate(alice)` and alice is not even a staker, then we will fail. Yeah you are right, defensive would produce too many false negatives here.

Reply:
On top of that, the defensives in testing are a bit too strict to work with. There are cases where we test that `validate(reaped_stash)` and ensure that we get a `NotStash` error, for example. And with defensive, we need to handle the panic instead of the returned error.

There are ways to handle this by breaking down and re-org the tests but I've come across this and wish that it would be easier to assert a defensive and runtime error at the same time and be able to proceed with the tests. I think this could be improved at the FRAME testing level, I'll think more about it.